src/main/java/net/minecraftforge/event/ModVersionsMismatchEvent.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.event;
+
+
+import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.fml.ModContainer;
+import net.minecraftforge.fml.ModList;
+import net.minecraftforge.fml.event.IModBusEvent;
+import net.minecraftforge.forgespi.language.IModInfo;
+import org.apache.maven.artifact.versioning.ArtifactVersion;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * <p>
+ * Fires when the mod loader is in the process of loading a world that was last saved
+ * with mod versions that differ from the currently-loaded versions. This can be used to
+ * enqueue work to run at a later point, such as multi-file migration of data.
+ * </p>
+ *
+ * <p>
+ * <b>Note that level and world information has not yet been fully loaded;</b> as such, it is
+ * unsafe to access server or level information during handling of this event.
+ * </p>
+ */
+public class ModVersionsMismatchEvent extends Event implements IModBusEvent
+{
+    /**
+     * The level being loaded. Useful for things like {@link net.minecraft.world.level.storage.DimensionDataStorage}
+     * to manage multiple files changing between mod versions.
+     */
+    private final LevelStorageSource.LevelDirectory levelDirectory;
+
+    /**
+     * A set of previously-known versions that have mismatched with the currently loaded versions.
+     */
+    private final Map<String, ArtifactVersion> previousVersions;
+
+    /**
+     * State values of which mods have specified that they have handled version mismatches.
+     */
+    private final Map<String, MismatchHandlingState> states;
+
+    /**
+     * Create an event instance to handle mod version mismatches.
+     *
+     * @param levelDirectory The directory of the level being loaded.
+     * @param previousVersions Map of mismatched mod ids and their previously-loaded version.
+     */
+    public ModVersionsMismatchEvent(LevelStorageSource.LevelDirectory levelDirectory, Map<String, ArtifactVersion> previousVersions)
```

Comment:
The javadoc can be omitted, and the event constructor should be annotated with `@ApiStatus.Internal`.